lib/zephyr/src/zephyr/plan.py

```diff
@@ -0,0 +1,996 @@
+    - Raw data (int, dict, str, etc.) for from_list/from_iterable pipelines
+    """
+
+    shard_idx: int
+    data: Any
+
+
+@dataclass
+class Map:
+    """Transform stream → stream via composed callable.
+
+    Attributes:
+        fn: Composed function that transforms an iterator to an iterator
+        requires_full_shard: True if any composed op needs full shard context
+            (e.g., MapShardOp). When True, chunk parallelism is disabled.
+    """
+
+    fn: Callable[[Iterator], Iterator]
+    requires_full_shard: bool = False
+
+
+@dataclass
+class Write:
+    """Write stream to file, return path."""
+
+    output_pattern: Callable[[int, int], str]  # (shard_idx, total_shards) → path
+    writer_type: str  # For chunk aggregation: "jsonl", "parquet", "binary", "levanter_cache"
+    skip_existing: bool = False
+    # Writer-specific parameters
+    levanter_metadata: dict | None = None
+    schema: Any = None  # For parquet
+    batch_size: int = 1000  # For parquet
+
+
+@dataclass
+class Scatter:
+    """Distribute items to output shards by key hash."""
+
+    key_fn: Callable[[Any], Any]  # item → key
+    num_output_shards: int
+
+
+@dataclass
+class Reduce:
+    """Merge sorted chunks and reduce per key."""
+
+    key_fn: Callable[[Any], Any]
+    reducer_fn: Callable[[Any, Iterator], Any]
+
+
+@dataclass
+class Fold:
+    """Reduce stream to single value."""
+
+    fn: Callable[[Iterator], Any]
+
+
+@dataclass
+class Reshard:
+    """Redistribute chunks across shards."""
+
+    num_shards: int
+
+
+@dataclass
+class Join:
+    """Join two sorted streams."""
+
+    fn: Callable[[Iterator, Iterator], Iterator]
+    right_plan: PhysicalPlan | None = None
+
+
+@dataclass
+class ForkChunks:
+    """Fork stream into N parallel chunk streams.
+
+    Child operations are applied in parallel, and merged as available.
+    """
+
+    target_chunks: int = DEFAULT_INTRA_SHARD_PARALLELISM
+    parallel_ops: list = field(default_factory=list)  # list[PhysicalOp]
+
+
+PhysicalOp = Map | Write | Scatter | Reduce | Fold | Reshard | Join | ForkChunks
+
+
+class StageType(Enum):
+    """Type of stage execution."""
+
+    WORKER = auto()  # Normal worker execution
+    RESHARD = auto()  # Redistribute chunks (no worker execution)
+
+
+def _map_gen(stream: Iterator, fn: Callable) -> Iterator:
+    for item in stream:
+        yield fn(item)
+
+
+def _filter_gen(stream: Iterator, predicate: Callable) -> Iterator:
+    for item in stream:
+        if predicate(item):
+            yield item
+
+
+def _flatmap_gen(stream: Iterator, fn: Callable) -> Iterator:
+    for item in stream:
+        yield from fn(item)
+
+
+def _reduce_gen(shard: Any, key_fn: Callable, reducer_fn: Callable) -> Iterator:
+    for key, items_iter in _merge_sorted_chunks(shard, key_fn):
+        yield reducer_fn(key, items_iter)
+
+
+def _select_gen(stream: Iterator, columns: tuple[str, ...]) -> Iterator:
+    cols_set = set(columns)
```

Comment:
Should we try to guarantee/preserve the order of the columns?

Reply:
i think it's okay this way since we're only using it for the dictionary generation?